### ISSUE TITLE
Migrations Bundles

### ISSUE THREAD

**Issue by bricelam** [dotnet/efcore#19693]
We currently have `dotnet ef migrations script` and `database update` to apply migrations. Unfortunately, they each have their shortcomings.

`migrations script` is pretty good at producing SQL scripts (there are some known issues with the scripts), but figuring out what to do with those scripts may not be obvious. Publish from Visual Studio wraps these scripts up in a web deployment package and uses MSDeploy to run them on the server. This works well for SQL Server on Windows, but nothing else.

`database update` is a lot better at applying migrations. It doesn't have any of the issues the SQL script does. Unfortunately, it requires the .NET Core SDK, your app's source code, and a direct connection to the production database. Having all of these things available where you need them is't always possible.

I propose we introduce a new `migrations bundle` command. This would create a self-contained EXE that you could use to apply migrations. It would behave the same as `database update` and let you specify the connection string. This would make SSH/Docker/PowerShell deployment a lot easier. You wouldn't need to copy your source files to a server inside the firewall and you wouldn't need to install the .NET Core SDK. Here is an example of what it might look like to use SSH:

```sh
dotnet ef migrations bundle
scp bundle me@myhost.com:./
ssh me@myhost.com "./bundle --connectionString ${DEPLOYMENT_CONNECTION_STRING}"
ssh me@myhost.com "rm bundle"
```

Taking this further, Visual Studio could do all of this for you like they do today on MSDeploy but when you're deploying to Docker. They could even create a new SSH deployment experience.

---

### TODO

- [x] Add `--output` to allow renaming the bundle
- [x] Match TargetFramework, ~~RuntimeIdentifier, and SelfContained~~ of startup project by default
  - ~~Add `--no-self-contained`~~
- [x] `--self-contained` doesn't work
  > error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false.
- [x] `--runtime` doesn't work between Windows and other OSes. We look for *.exe based on the *current* runtime instead of the *target* runtime.
- [x] Add automated tests
- [x] Add `--force` to overwrite an existing bundle #25271
- [x] Refine command output #25274
- [x] `--configuration` and `--runtime` optoins are repeated in help

**Comment by tibitoth** [comment]
@bricelam Why not use dacpac? I have not seen any better approach for MSSQL.

We try to embrace sqlproj (in this sense it's DB first) + reverse engineer code first on every change + dacpac migrations.

**Comment by bricelam** [comment]
dacpac is perfect for DB-first; keep using it there. However, it's fundamentally different from Migrations which are designed for a code-first workflow. Also, dacpac is SQL Server-only and a big part of these bundles is to improve the experience when using other databases like PostgreSQL, MySQL, and Oracle.

Fun fact, the first prototype of Migrations was based on dacpac.

**Comment by ErikEJ** [comment]
FWIW, "EF Core Power Tools" supports Reverse Engineering directly from a dacpac / SQL Database porjecy (.sqlproj)

**Comment by leealexander** [comment]
> migrations script is pretty good at producing SQL scripts (there are some known issues with the scripts)

Could you elaborate on the issues?

>database update is a lot better at applying migrations.

In what way(s) is update better?

I take it the SQL isn't the same if I was to look at trace logs of the two different update methods?

**Comment by bricelam** [comment]
The SQL is the same, the differences (and issues) stem from things like when transactions begin and end, the if blocks (some statements aren’t allowed inside), and how the script is parsed (some objects are created at the beginning of the script but the parser/compiler errors because later statements reference them.

**Comment by bricelam** [comment]
Items to address issues with the script: #7681 #12911 #14746 #19740

**Comment by ajcvickers** [comment]
Team note: make sure to consider how bundles interact with continuous deployment systems. See #9841.

**Comment by ajcvickers** [comment]
Unfortunately, while we have made good progress in understanding the space here, we are deferring release of the migrations bundles feature until after the EF Core 5.0 release to provide more time for feedback. The current plan is to get something into preview near the beginning of the EF Core 6.0 milestone so that we have enough time for feedback, integration, and testing before committing.

**Comment by ivanmonteiro** [comment]
I'm not sure if it is the right place, but here are my considerations:
It would be interesting if the migrations bundle could be built and run on both Windows/Linux platforms. 

Currently I'm developing using a Windows machine and deploying to a Linux server. Also, my CI workflow uses Linux machines under the covers. So being able to choose the target platform would be great.

**Comment by JohnnySixStrings** [comment]
This is currently an issue I'm interested in we currently work with multiple databases including postgres, mysql, sql server. 
Using azure pipelines we have not found a suitable way to deploy migrations since migration script will not work with postgres.

**Comment by roji** [comment]
@JohnnySixStrings migration scripts generally work fine with PostgreSQL - I'm guessing you're running into issues with maintaining migrations with multiple providers ([see this doc page](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/providers?tabs=dotnet-core-cli)).

If you're encountering specific issues with PostgreSQL migrations, that probably isn't related to this issue - can you please open an issue on https://github.com/npgsql/efcore.pg?

**Comment by rhyswat** [comment]
A common use case for me is that the production database is not accessible from my workstation - not directly, and not over a tunnel either - typically due to clients' on-prem requirements.

So the proposed bundle -> copy -> run -> cleanup workflow would work well in these cases.

One thing to consider is that the putative bundle executable needs to be runnable on a different platform to the one where it was built. For example, my use case might require building the bundle on an Ubuntu box then copying it and running it against an MS SQL database on a Windows Server box.

I realise that we can target runtimes when building our applications, so I wonder whether a) this is necessary and b) whether this is possible for bundle executables.

**Comment by ivanmonteiro** [comment]
Currently my use case is using docker for deployment with heavy use of CI/CD platforms such as Circle CI, Github Actions and others. They might use Linux OS as underlying infrastructure. So multi-platform targeting fits my use case. Also, my client's target OS or existing infrastructure might be Windows or Linux.

With migrations bundle ideally It would be possible to also _run_ the migrations using the [runtime docker image](https://hub.docker.com/_/microsoft-dotnet-runtime/) instead of the [sdk image](https://hub.docker.com/_/microsoft-dotnet-sdk/) which is bigger and not ideal for final stages of docker containers. 

I see no problem using the dotnet core sdk image for the _build_ stage of the bundle but actually running the migrations bundle should be using the more lightweight [runtime docker image instead](https://hub.docker.com/_/microsoft-dotnet-runtime/).

**Comment by bricelam** [comment]
### 📢 Status update

I got an end-to-end working. There are still a lot of details to sort out before sending a PR...

![Bundles](https://user-images.githubusercontent.com/1226749/124196426-4187fe80-da81-11eb-8a80-3de44adb09bb.gif)

**Comment by bricelam** [comment]
An initial implementation was just merged. It'll be in Preivew 7 (August).

You can try it out using the [daily builds](https://github.com/dotnet/efcore/blob/main/docs/DailyBuilds.md):

```sh
dotnet tool update -g dotnet-ef --version 6.0.0-* --add-source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json
dotnet ef migrations bundle
./bundle
```

Or from Visual Studio:

```ps1
Bundle-Migration
./bundle.exe
```

### TODO

- [ ] Add `--output` to allow renaming the bundle
- [ ] Match TargetFramework, RuntimeIdentifier, and SelfContained of startup project by default
  - Add `--no-self-contained`
- [ ] `--self-contained` doesn't work
  > error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false.
- [ ] `--runtime` doesn't work between Windows and other OSes. We look for *.exe based on the *current* runtime instead of the *target* runtime.
- [ ] Add automated tests

**Comment by ErikEJ** [comment]
Cannot see in daily feed yet?

**Comment by julielerman** [comment]
Hey @bricelam ...should this work with SQLite yet? I'm targeting 6.0.0-rc.1.21409.2 for packages and the tools. Getting 

`Method 'AppendIdentityWhereCondition' in type 'Microsoft.EntityFrameworkCore.Sqlite.Update.Internal.SqliteUpdateSqlGenerator' from assembly 'Microsoft.EntityFrameworkCore.Sqlite, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.`

If that's new to you then maybe it's related to my model and I can dig further. THanks

**Comment by julielerman** [comment]
Note: I removed the owned entity from the model, trashed original migrations and created a new migration from scratch. Same error when trying to execute the new bundle so it wasn't the owned entity. :)

**Comment by roji** [comment]
@julielerman this is most likely as a result of mixing EF Core versions - can you check in your csproj that all versions are exactly the same etc.?

**Comment by julielerman** [comment]
@roji Hi there. 
Only package refs are 6.0.0-*

I am targetting daily builds with nuget.config 

```
<packageSources>
        <add key="dotnet6" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json" />
        <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
    </packageSources>
```

In projects.assets.json all efcore packages are the same version:

"Microsoft.EntityFrameworkCore.Sqlite/6.0.0-rc.1.21409.2": {
      
Microsoft.EntityFrameworkCore.Relational/6.0.0-rc.1.21409.2": {
    "Microsoft.EntityFrameworkCore.Design/6.0.0-rc.1.21409.2": {

And I installed the same version of the efcore tools as per Brice's note above.

I'm going to test against sql server now (just had to get my container running :) )

**Comment by julielerman** [comment]
same with sql server. Ok...so it's me.  I just don't know what... :)

**Comment by ErikEJ** [comment]
Please share your .csproj

**Comment by julielerman** [comment]
not much in there:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="6.0.0-*" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="6.0.0-*" />
   <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlserver" Version="6.0.0-*" />
  
  </ItemGroup>

</Project>
```

**Comment by julielerman** [comment]
GOT IT! Even though 6.0.0.-* was giving me the correct versions (as per assets file), specifically targeting brought me success

    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="6.0.0-rc.1.21409.2" />
   <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlserver" Version="6.0.0-rc.1.21409.2" />

**Comment by julielerman** [comment]
update on my SQLite bundle problem (Unable to load shared library e_sqlite3 or one of its dependencies)
Failing on TWO MacBooks.  I spent a bunch of hours trying to find a solution. (Reminder: that worked when I targeted SQL Server).
Same project ...success on Windows.
Gah!
// @bricelam (sorry)

**Comment by gunniwho** [comment]
Is it possible to provide the connection string via the file system? Something like `bundle --connection-file /vault/secrets/connection`

**Comment by bricelam** [comment]
@gunniwho Yes, like `dotnet ef`, the bundle honors response files.
```sh
bundle --connection-file @/vault/secrets/connection
```

**Comment by EricStG** [comment]
Would it be possible to add an option to the bundle to preview the scripts (`--script` or `--dryrun` or something)? Ideally, we'd want a manual approval step in our pipeline after human eyeballs review the output.
I know this can be generated from the ef core tools, but ideally we'd want to use the bundle's artifact on its own

**Comment by lohithgn** [comment]
@bricelam Awesome addition to EF. Where can I find official documentation on migration bundles ?

**Comment by roji** [comment]
@lohithgn docs aren't yet available as this feature is just now being finished. They should appear by the time EF Core 6.0 ships, and will be available under the [migration section of the docs](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/?tabs=dotnet-core-cli).

**Comment by ryanrichard19** [comment]
Attempting to run Bundle-Migration on the Packege-Manager Console with latest VS Preview and getting the following error 
Bundle-Migration
Build started...
Build succeeded.
Building bundle...
Build failed. Use --verbose to see errors.

The way to create the migration and build it works fine for me. 

Here is my project metadata 
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.0-preview.7.21378.4" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="6.0.0-preview.7.21378.4" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.0-preview.7.21378.4" />
  </ItemGroup>

</Project>

Hope this helps

**Comment by julielerman** [comment]
@ryanrichard19 can you run it with --verbose and share any errors that reveals?

**Comment by ryanrichard19** [comment]
When I try that I  get this @julielerman 
Bundle-Migration --verbose
Bundle-Migration : A positional parameter cannot be found that accepts argument '--verbose'.
At line:1 char:1
+ Bundle-Migration --verbose
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Bundle-Migration], ParameterBindingException
    + FullyQualifiedErrorId : PositionalParameterNotFound,Bundle-Migration

**Comment by ErikEJ** [comment]
Please try running with -Verbose

**Comment by ryanrichard19** [comment]
Thanks @ErikEJ now I get this 
PM> Bundle-Migration -Verbose
Using project 'MarketingIncentive.Data'.
Using startup project 'MarketingIncentive.API'.
Build started...
Build failed.

**Comment by ErikEJ** [comment]
And both these projects can build in VS?

**Comment by ryanrichard19** [comment]
Yes I can run migrations the other way and that works for me. Could it be due to me having the Data project separated?

**Comment by ErikEJ** [comment]
Does your entire solution build without errors??

**Comment by julielerman** [comment]
Since I'm working on my macbook, I'm curious if you get the same problem if you go to the command line and run  the CLI version (dotnet ef migrations bundle). My small demo isn't in separate projects but I can try that.

**Comment by ryanrichard19** [comment]
Should I be in the .Data project folder when I run the CLI command?

**Comment by ryanrichard19** [comment]
from the .data folder
❯ dotnet ef migrations bundle
Build started...
Build succeeded.
Specify --help for a list of available options and commands.
Unrecognized command or argument 'bundle'
⨯ mrrar@DESKTOP-0VK5K1M  ~\Development\MarketingIncentive\MarketingIncentive.Data   main ≣ +2 ~26 -1 ! 

**Comment by ryanrichard19** [comment]
Just ran the following 
 dotnet tool update --global dotnet-ef --version 6.0.0-preview.7.21378.4
Tool 'dotnet-ef' was successfully updated from version '3.1.5' to version '6.0.0-preview.7.21378.4'.
mrrar@DESKTOP-0VK5K1M  ~\Development\MarketingIncentive\MarketingIncentive.Data   main ≣ +2 ~26 -1 !                                                                                                [17:10]
❯ dotnet ef migrations bundle
Build started...
Build succeeded.
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.EntityFrameworkCore.Design, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
File name: 'Microsoft.EntityFrameworkCore.Design, Culture=neutral, PublicKeyToken=null'
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark, IntPtr pPrivHostBinder)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.Load(String assemblyString)
   at Microsoft.EntityFrameworkCore.Tools.Commands.ProjectCommandBase.get_EFCoreVersion()
   at Microsoft.EntityFrameworkCore.Tools.Commands.MigrationsBundleCommand.Execute(String[] args)
   at Microsoft.DotNet.Cli.CommandLine.CommandLineApplication.Execute(String[] args)
   at Microsoft.EntityFrameworkCore.Tools.Program.Main(String[] args)


Could not load file or assembly 'Microsoft.EntityFrameworkCore.Design, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

**Comment by ErikEJ** [comment]
Can you share your .csproj?

**Comment by julielerman** [comment]
sorry I don' t mean to lead you down a completely different set of problems. Follow Erik's lead in VS /Windows. Also, (on mac) I just separated my demo into three folders: domain (library), data (library) and app (console). I'm using SQLite and nightly builds and bundle worked as expected.

**Comment by ryanrichard19** [comment]
@julielerman no worries. Amazing trying to figure this out with you and @ErikEJ.

@ErikEJ I can share this proj as it is in a private repo for a client. Will set up a new public repo to see if I get this error. Thanks all for your help.

**Comment by ryanrichard19** [comment]
So I added the ef design package to my project 

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.0-preview.7.21378.4" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="6.0.0-preview.7.21378.4">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="6.0.0-preview.7.21378.4" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.0-preview.7.21378.4" />
  </ItemGroup>

</Project>

and rerun the command in Package Console manger and i get this now

PM> Bundle-Migration -Verbose
Using project 'MarketingIncentive.Data'.
Using startup project 'MarketingIncentive.API'.
Build started...
Build succeeded.
C:\Program Files\dotnet\dotnet.exe exec --depsfile C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.API.deps.json --additionalprobingpath C:\Users\mrrar\.nuget\packages --additionalprobingpath "C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages" --additionalprobingpath "C:\Program Files (x86)\Microsoft\Xamarin\NuGet" --additionalprobingpath "C:\Program Files\dotnet\sdk\NuGetFallbackFolder" --runtimeconfig C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.API.runtimeconfig.json C:\Users\mrrar\.nuget\packages\microsoft.entityframeworkcore.tools\6.0.0-preview.7.21378.4\tools\netcoreapp2.0\any\ef.dll migrations bundle --verbose --no-color --prefix-output --assembly C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll --project C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\MarketingIncentive.Data.csproj --startup-assembly C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.API.dll --startup-project C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj --project-dir C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\ --language C# --working-dir C:\Users\mrrar\Development\MarketingIncentive --root-namespace MarketingIncentive.Data
Using assembly 'MarketingIncentive.Data'.
Using startup assembly 'MarketingIncentive.API'.
Using application base 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0'.
Using working directory 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API'.
Using root namespace 'MarketingIncentive.Data'.
Using project directory 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'MarketingIncentive.API'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Using application service provider from Microsoft.Extensions.Hosting.
Found DbContext 'MarketingIncentiveDBContext'.
Finding DbContext classes in the project...
Using context 'MarketingIncentiveDBContext'.
Building bundle...
dotnet publish --output C:\Users\mrrar\AppData\Local\Temp\etc5av4u.vrh\publish
Microsoft (R) Build Engine version 17.0.0-preview-21378-03+d592862ed for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\Users\mrrar\AppData\Local\Temp\etc5av4u.vrh\bundle.csproj (in 319 ms).
  2 of 3 projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  MarketingIncentive.Data -> C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 2 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 3 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 4 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 5 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 6 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 7 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 8 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 9 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): warning MSB3026: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Beginning retry 10 in 1000ms. The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): error MSB3027: Could not copy "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". Exceeded retry count of 10. Failed.  [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Microsoft.Common.CurrentVersion.targets(4685,5): error MSB3021: Unable to copy file "C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.Data\bin\Debug\net6.0\MarketingIncentive.Data.dll" to "bin\Debug\net6.0\MarketingIncentive.Data.dll". The process cannot access the file 'C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\bin\Debug\net6.0\MarketingIncentive.Data.dll' because it is being used by another process. [C:\Users\mrrar\Development\MarketingIncentive\MarketingIncentive.API\MarketingIncentive.API.csproj]
Microsoft.EntityFrameworkCore.Tools.CommandException: Build failed. Use --verbose to see errors.
   at Microsoft.EntityFrameworkCore.Tools.Commands.MigrationsBundleCommand.Execute(String[] args) in ef.dll:token 0x60001a9+0x23b
   at Microsoft.EntityFrameworkCore.Tools.Commands.CommandBase.<>c__DisplayClass0_0.<Configure>b__0(String[] args) in ef.dll:token 0x6000205+0x3b
   at Microsoft.DotNet.Cli.CommandLine.CommandLineApplication.Execute(String[] args) in ef.dll:token 0x6000053+0x100
   at Microsoft.EntityFrameworkCore.Tools.Program.Main(String[] args) in ef.dll:token 0x60000cf+0x2e
Build failed. Use --verbose to see errors.

**Comment by ryanrichard19** [comment]
Just restarted my PC and retried and get the same error.

**Comment by bricelam** [comment]
Hmm, it looks like something is breaking the [incremental build](https://docs.microsoft.com/visualstudio/msbuild/incremental-builds) causing a full rebuild of the startup project. Do you have any custom targets?

**Comment by ryanrichard19** [comment]
No I dont. Pretty much standard entity framework and dbcontext setup

**Comment by bricelam** [comment]
I'm able to repro this on the latest bits. `dotnet ef` works but VS fails. Filing a new issue to follow up: #25555

**Comment by julielerman** [comment]
Yay @ryanrichard19 for uncovering it! :)

**Comment by marcusturewicz** [comment]
I may have missed this, but how to migrations bundles compare with idempotent sql scripts? For example, if I currently have this:

```
dotnet ef migrations script -i -p src -o publish-artifacts/sql/script.sql
```
Then deploy the script with `Azure/sql-deploy` action.

If I change this to:

```
dotnet ef migrations bundle -p src
./bundle --connection [my-connection-string]
```

will it still work in the same way?

**Comment by julielerman** [comment]
I had the same question and just tested it out last night since you can't see what's in the bundle. (Hmm I think we should have a way to verify bundle contents ... black box can be scary) 

It seems to contain all of the migrations and runs just the needed ones. 

For my test, I had two migrations and built two bundles: one after creating the first migration and another after creating the second. I ran the first with no db existing and it created the db and ran the single migration .  I ran the second and it ran only the second migration. Then I deleted the db and ran the second and it created the db and ran both scripts.

**Comment by julielerman** [comment]
@bricelam Shall I create an issue requesting a way to discover what migrations are included in a particular bundle?

**Comment by zeshan321** [comment]
Is there a way to revert migrations with the bundle?

**Comment by jaliyaudagedara** [comment]
What do you guys think of adding an argument to support running the migrations in a transaction? I think that's a huge pain point (at least for me, I believe there are a lot of other devs as well).

I just tried 2 migrations, the first one should pass, and the second one introduces a unique constraint that should fail on purpose. When I run the bundle.exe, it throws an error (as expected), but the first migration is applied. This causes the database to be in an inconsistent state.

Most of the time I use `migrations scripts` and in the pipeline, I am manually wrapping the script execution in a transaction. So either all gets executed or non gets executed. 

What do you guys think?

**Comment by roji** [comment]
@jaliyaudagedara see #22616 for relevant discussion, though that issue discusses both script and non-script ways of applying migrations.

**Comment by jaliyaudagedara** [comment]
Thanks @roji!

**Comment by bricelam** [comment]
> Shall I create an issue requesting a way to discover what migrations are included in a particular bundle?

Today the bundle is effectively just `dotnet ef database update`, but I've been thinking it would be helpful to include gestures for these commands too in the bundle:

* `dotnet ef dbcontext info` (or even just a way to see the DbContext type of the bundle in --help)
* `dotnet ef migrations list`
* `dotnet ef migrations script` (altough this seems much less important since you can just generate a script instead of a bundle in the first place)

**Comment by ajcvickers** [comment]
@bricelam Yes!

**Comment by ivanmonteiro** [comment]
I might be missing something, but I tried to create the bundle inside a (linux) docker container. 

**The bundle file is generated but I cannot run it.**

This does not seem to be a permission problem, because the user is root and the bunde has permission:
```
/app # whoami
root
/app # ls -lah /app/
-rwxr--r--    1 root     root        3.6M Apr 20 10:26 Swashbuckle.AspNetCore.SwaggerUI.dll
-rwxr-xr-x    1 root     root        1.1K Aug 26 23:45 appsettings.Development.json
-rwxr-xr-x    1 root     root        1.1K Aug 26 19:18 appsettings.Production.json
-rwxr-xr-x    1 root     root        1.3K Aug 26 20:53 appsettings.json
-rwxr-xr-x    1 root     root       33.7M Sep  7 23:07 bundle
```

I tried to run the bundle in two ways:
With `./bundle` returns:
```
./bundle
sh: ./bundle: not found
```
and with the command `dotnet bundle` returns:
```
# dotnet bundle
Could not execute because the application was not found or a compatible .NET SDK is not installed.
Possible reasons for this include:
  * You intended to execute a .NET program:
      The application 'bundle' does not exist.
  * You intended to execute a .NET SDK command:
      It was not possible to find any installed .NET SDKs.
      Install a .NET SDK from:
        https://aka.ms/dotnet-download
```

Note that I'm using the **mcr.microsoft.com/dotnet/sdk:6.0-focal** image as the builder image to build the application and the bundle. 
And **mcr.microsoft.com/dotnet/aspnet:6.0-alpine** to run the application and to execute the bundle file.
Here is my Dockerfile:
```
FROM mcr.microsoft.com/dotnet/sdk:6.0-focal AS build

# Add dotnet-ef tool to create the migrations bundle
RUN dotnet tool install --global dotnet-ef --version 6.0.0-preview.7.21378.4
ENV PATH="${PATH}:/root/.dotnet/tools"

# Restore dependencies of .net core projects
COPY src/*/*.csproj ./src/
COPY Directory.Build.targets Directory.Build.targets
RUN for file in $(ls src/*.csproj); do mkdir -p ${file%.*} && mv $file ${file%.*}; done
RUN dotnet restore "src/BackendApp/BackendApp.csproj"

# Copy everything else and build app
COPY . ./
WORKDIR src/BackendApp
RUN dotnet publish "BackendApp.csproj" -c Release -o /app/out/
RUN dotnet ef migrations bundle --verbose && mv bundle /app/out/

# Final stage using dotnet runtime (alpine)
FROM mcr.microsoft.com/dotnet/aspnet:6.0-alpine AS runtime
WORKDIR /app
EXPOSE 80
ENV PORT=80
COPY --from=build /app/out .

ENV ASPNETCORE_ENVIRONMENT=Production

CMD ASPNETCORE_URLS=http://*:$PORT dotnet backendapp.dll

```

I wonder if I'm missing something 🤔

**Comment by bricelam** [comment]
@ivanmonteiro Can you try again on RC1? If it's still an issue, please submit a new issue.

**Comment by M2xAy** [comment]
Hi Community
We have also this error because it is being used by another process. on bundleing on ef core 6

Any updates ?
Best regards from zurich

**Comment by meronz** [comment]
> I might be missing something, but I tried to create the bundle inside a (linux) docker container.
> 
> **The bundle file is generated but I cannot run it.**
> 
> This does not seem to be a permission problem, because the user is root and the bunde has permission:
> 
> ```
> /app # whoami
> root
> /app # ls -lah /app/
> -rwxr--r--    1 root     root        3.6M Apr 20 10:26 Swashbuckle.AspNetCore.SwaggerUI.dll
> -rwxr-xr-x    1 root     root        1.1K Aug 26 23:45 appsettings.Development.json
> -rwxr-xr-x    1 root     root        1.1K Aug 26 19:18 appsettings.Production.json
> -rwxr-xr-x    1 root     root        1.3K Aug 26 20:53 appsettings.json
> -rwxr-xr-x    1 root     root       33.7M Sep  7 23:07 bundle
> ```
> 
> I tried to run the bundle in two ways: With `./bundle` returns:
> 
> ```
> ./bundle
> sh: ./bundle: not found
> ```
> 
> and with the command `dotnet bundle` returns:
> 
> ```
> # dotnet bundle
> Could not execute because the application was not found or a compatible .NET SDK is not installed.
> Possible reasons for this include:
>   * You intended to execute a .NET program:
>       The application 'bundle' does not exist.
>   * You intended to execute a .NET SDK command:
>       It was not possible to find any installed .NET SDKs.
>       Install a .NET SDK from:
>         https://aka.ms/dotnet-download
> ```
> 
> Note that I'm using the **mcr.microsoft.com/dotnet/sdk:6.0-focal** image as the builder image to build the application and the bundle. And **mcr.microsoft.com/dotnet/aspnet:6.0-alpine** to run the application and to execute the bundle file. Here is my Dockerfile:
> 
> ```
> FROM mcr.microsoft.com/dotnet/sdk:6.0-focal AS build
> 
> # Add dotnet-ef tool to create the migrations bundle
> RUN dotnet tool install --global dotnet-ef --version 6.0.0-preview.7.21378.4
> ENV PATH="${PATH}:/root/.dotnet/tools"
> 
> # Restore dependencies of .net core projects
> COPY src/*/*.csproj ./src/
> COPY Directory.Build.targets Directory.Build.targets
> RUN for file in $(ls src/*.csproj); do mkdir -p ${file%.*} && mv $file ${file%.*}; done
> RUN dotnet restore "src/BackendApp/BackendApp.csproj"
> 
> # Copy everything else and build app
> COPY . ./
> WORKDIR src/BackendApp
> RUN dotnet publish "BackendApp.csproj" -c Release -o /app/out/
> RUN dotnet ef migrations bundle --verbose && mv bundle /app/out/
> 
> # Final stage using dotnet runtime (alpine)
> FROM mcr.microsoft.com/dotnet/aspnet:6.0-alpine AS runtime
> WORKDIR /app
> EXPOSE 80
> ENV PORT=80
> COPY --from=build /app/out .
> 
> ENV ASPNETCORE_ENVIRONMENT=Production
> 
> CMD ASPNETCORE_URLS=http://*:$PORT dotnet backendapp.dll
> ```
> 
> I wonder if I'm missing something thinking

```
RUN apk add --no-cache gcompat
```
Works for me.

**Comment by ivanmonteiro** [comment]
> @ivanmonteiro
> 
> > I wonder if I'm missing something thinking
> 
> ```
> RUN apk add --no-cache gcompat
> ```
> 
> Works for me.

I ended up using the alpine docker image in both build stage and the final stage of the Dockerfile and that solved the ` ./bundle: not found` problem for me.

**Comment by pinkfloydx33** [comment]
According to the documentation, it looks like you can pass the target migration name to the bundle. Can this be used to downgrade the migrations?

In other words, suppose we deploy our database and application in the same pipeline. We deploy the database successfully but the application deployment fails for whatever reason. Could we use the same bundle to undo the migrations (for ex. in an on-failure step) assuming we know the previous migration name? If yes, is there a way to find out what the most recent migration was so that we know what to roll back to (either using bundle or direct sql)?

**Comment by sabbadino** [comment]
why when i run "dotnet ef migrations bundle" , it tries to connect to the db using appsettings cnstring ?  (and i get a connection error) .. I expected that this command would just bundle migrations

**Comment by bricelam** [comment]
@sabbadino It shouldn't connect to the database. Do you have something in your Program.CreateHostBuilder (or Main for minimal API projects) or Startup.ConfigureServices that connects to the database?

**Comment by BenjaminAbt** [comment]
Any hint (even if not supported today) how to use bundles with Azure DevOps and Managed Identity on Azure SQL?
Username:Password in Connectionstring is meh.

Or fallback to ef migration script (or dacpac) and Azure SQL Deployment Task with ServicePrincipal, if we wanna have secure deployments?

**Comment by ErikEJ** [comment]
@BenjaminAbt There are various options: https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/azure-active-directory-authentication?view=sql-server-ver15 - but not sure I understand your exact requirement - what does the managed identity represent?

**Comment by pinkfloydx33** [comment]
You need to make sure your deployment agent VM has an MI/SP assigned to it. Then create a contained user in your azure sql db. Then for a connection string, you don't need user/password just `Authentication=Active Directory Default`. Nothing special other than that and it should just work. I do similar for both Postgres and Azure SQL

**Comment by BenjaminAbt** [comment]
Thanks for your replies.

> Then for a connection string, you don't need user/password just Authentication=Active Directory Default

I am doing that. I was afraid that this would not work conceptually - but then there seems to be a bug in my setup.
Thank you very much for the feedback.

**Comment by ErikEJ** [comment]
@BenjaminAbt for Active Directory Defsult  make sure to use SqlClient 3.0 or later

**Comment by sabbadino** [comment]
@bricelam ,sure i have , it's my app that does to connect to the db, I really don't get the question ? Am i missing something ? Is this bundle supposed to generate the scripts to be executed at build time ? I expected the bundle would bring all migration in it and evaluate which ones to run at deploy time ..

**Comment by pinkfloydx33** [comment]
> Thanks for your replies.
> 
> > Then for a connection string, you don't need user/password just Authentication=Active Directory Default
> 
> I am doing that. I was afraid that this would not work conceptually - but then there seems to be a bug in my setup. Thank you very much for the feedback.

If you are using Azure devops, you may also want to try running the Azure CLI task and using the service principal assigned to that, executing the bundle as the SP. Just make sure that SP has its own "contained user" in your db. That should also work. If you're using am older version of Ms.Data.SqlClient you may need to use "Active Directory Managed Identity" or one of the other flavors, though they require the appid for the user name

**Comment by bricelam** [comment]
@sabbadino bundles currently work the same way that `dotnet ef database update` does. It needs to build your application service provider in order to discover and instantiate the DbContext. It does this by calling Program.CreateHostBuilder. So every time you run the bundle, that method is executed.

**Comment by sabbadino** [comment]
So in opinion that is totally useless.. when i build i want to make a single  bundle that will evaluate what updates to apply at deploy time 

If i got it correctly, the way it works is 
1) i should make a bundle for each db i want to target (cicd, test, prd)
2) build agent must reach any of the db of the different environments.

Not a real world situation

**Comment by pinkfloydx33** [comment]
It builds an Exe that you pass a connection string to and it evaluates what migrations are missing and applies them. It can also rollback or only migrate to a specific migration (via commandline).

I build a bundle in CI and attach the single executable as an artifact which I then use during CD without having to build anything again. Not sure what you're talking about. They work as you expect them to

**Comment by ivanmonteiro** [comment]
> 1. i should make a bundle for each db i want to target (cicd, test, prd)

No. Ideally you build the bundle only once. Then you can pass to the bundle.exe either a connection string or it gets the connection string from the  appsettings.{Environment}.json file. 

> 2. build agent must reach any of the db of the different environments.

Also, from my experience with the migrations bundle, when building the migrations bundle no connection to the database is needed in order to create the bundle.exe. 
Now, the bundle.exe needs to connect to the database in order to apply the migrations, of course.
As mentioned by @julielerman you can generate a migration script (`dotnet ef migrations script --output migrations.sql`) instead of the bundle if it suits your needs.

**Comment by sabbadino** [comment]
Hi, i Will double check .. if the problem shows up actually, i will post the call stack

**Comment by chinthakaherath** [comment]
Is there any guide or document on how to run efbundle.exe in the azure release pipeline?

**Comment by chinthakaherath** [comment]
Is there any guide or document on how to run efbundle.exe in the azure release pipeline?

**Comment by SirMrJames** [comment]
> Is there any guide or document on how to run efbundle.exe in the azure release pipeline?

I'm having the same issue, did you ever figure that out?

**Comment by ErikEJ** [comment]
Just run a script with [the command line](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/applying?tabs=dotnet-core-cli#efbundle) in you yaml pipeline:

```ps
- pwsh:  .\efbundle.exe --connection 'Data Source=(local)\MSSQLSERVER;Initial Catalog=Blogging;User ID=myUsername;Password=myPassword'
  failOnStderr:  true
```

**Comment by fredriksjoholm** [comment]
Is there any way to run bundles in a distributed settings while preventing two bundles from applying changes to the database simultaneously? What I'm trying to do is to apply my migrations from an init container in a Kubernetes cluster. I've looked into using Kubernetes jobs in order to ensure that each bundle only runs once, but this approach seems to have its own drawbacks as well.

**Comment by roji** [comment]
@fredriksjoholm pervent concurrent migrations isn't something that EF does - you'll have to take care of it outside, where you're applying the migrations (e.g. in Kubernetes). Note that in general, applying migrations is seen as part of deploying a new version of your application; this is already a process that should happen once and in a controlled fashion (you don't want to have two different application deployments racing concurrently!).

For context, EF could try to make sure that migrations aren't executed concurrently by e.g. taking a lock on the database's migrations history table. The main problem with that is that migrations may involve multiple transactions (e.g. since SQL Server doesn't support various DDL statements in migrations), and so the lock would be released in the middle of the transaction. But once again, even regardless of the technical difficulties, this is something that should be handled at a higher level, in your deployment pipeline.

**Comment by roji** [comment]
> Even in a situation where a migration was done in a CD pipeline there are no guarantees that the CD pipeline is going to run once. A couple of providers will allow triggering of multiple pipelines for the same commit, and most others will default to running the same pipeline in parallel for different commits. So it seems that the risk concurrency risk is present no mater where it is run.

I may be misunderstanding you, but if multiple *deployments* really are executing concurrently, wouldn't that mean that e.g. an actual production server might end up with mixed application files from two versions, as files are being copied from both onto it?
It's hard for me to see how this could be workable: the actual act of copying the files to production - and executing migrations - seems like it must be protected by some sort of concurrency control seems necessary at the deployment stage.

But maybe I'm misunderstanding what you're describing - if so, more detail/context could help.

> [...] I would think that it would be preferable to maybe right an official guide for how one would solve this instead of hoping that the first result gives the correct answer. [...]

I don't personally think the official EF documentation needs to address all possible scenarios with all possible technologies - there are many cluster and deployment technologies out there, and practically speaking we couldn't cover them all (nor would we be quailfied to do so). I don't see the problem with external documentation and blogs addressing scenarios such as this, especially since there's lots of diversity in exactly how people configure their clusters and deployment...

**Comment by phillebaba** [comment]
@roji I guess its difficult to explain without digging too deep into Kubernetes. I do agree with you that attempting to document every single deployment solution is not feasible. The issue is that the recommended solution would not work at all for other deployment solutions. Would a compromise be to collect a set of resources to point end users to known working solutions for different platforms?

**Comment by OskarKlintrot** [comment]
@phillebaba I don't understand how this has anything to do with EF really? Isn't it down to each CI/CD-solution to document? There're more tools than just EF bundle that you only want to run once pre- or post-deployment so it would make more sense if the CI/CD-provider documents how to do it rather than each individual tool. As an example, if I deploy db-changes (which there is a plethora of tools to choose from, [DbUp](https://dbup.github.io/) is quite popular) I would read in the Azure DevOps or GitHub Actions (my providers) docs to understand how I would make sure that if one deployment is already running then the next in line waits for that one to finish or get cancelled before the next deployment runs. I don't see why I would look in the EF documentation for how to use GitHub Actions (as an example)?

**Comment by roji** [comment]
@phillebaba what kind of answer or capability do you think EF could provide to help with this? At the end of the day, EF can only produce a migration bundle or SQL script, but it's entirely up to the user's environment how those will get applied/executed, and from where. Some very concrete suggestions would help understand what you have in mind.

**Comment by phillebaba** [comment]
Honestly adding a DB lock before running the migration would solve these issues. Technically this could be done by developer beforehand but I do not think that it should be the developers responsibility

There are tools like Liquibase that already do this and work without any issue with init containers in Kubernetes.
https://www.liquibase.com/blog/using-liquibase-in-kubernetes

**Comment by roji** [comment]
@phillebaba I addressed the issues with that [here](https://github.com/dotnet/efcore/issues/19693#issuecomment-1141857340).

**Comment by ElanHasson** [comment]
Is there a way to make this work with `dotnet pack`?

**Comment by roji** [comment]
@ElanHasson can you provide more context on what you want to make work with `dotnet pack`?

**Comment by ElanHasson** [comment]
> @ElanHasson can you provide more context on what you want to make work with `dotnet pack`?


I wrote pack, but I was thinking it should have been publish.

I was curious if dotnet publish could be used to run the bundler instead of the normal behavior.

Just trying to avoid modifying build scripts :)

**Comment by ErikEJ** [comment]
"Normal behaviour" IS dotnet publish 😄  https://github.com/ErikEJ/EFCorePowerTools/issues/1078#issuecomment-962194993

**Comment by ElanHasson** [comment]
Ahh I see. This is a bit more than I need @ErikEJ  but thank you. 
I'm using EF only for migrations, in my https://github.com/dotnet/orleans project via https://github.com/web-scheduler/web-scheduler/pull/159

**Comment by Gary-Elphick-Star** [comment]
Are migrations bundles fully supported now? If is, is there an Azure Pipelines deployment task specifically for it (to Azure SQL)?
Currently we use SqlAzureDacpacDeployment@1 to run migrations scripts on some of our environments.

**Comment by BenjaminAbt** [comment]
@Gary-Elphick-Star due to the authentication (Managed Identity) you have to run them in an Azure CLI task.

```yaml
- task: AzureCLI@2
  displayName: Azure SQL Schema Deployment
  inputs:
    azureSubscription: ${{ parameters.azureConnectionName }}
    scriptType: pscore
    scriptLocation: inlineScript
    inlineScript: |
      ${{ parameters.sqlServerDatabaseBundlePath}}\${{ parameters.sqlServerDatabaseBundleName }} `
        --connection 'Server=${{ variables.SqlServer }};Database=${{ variables.SqlDatabase }};Authentication=Active Directory Default'
```


They work great.
There is just no documentation how to run them on a Linux Agent.

**Comment by Gary-Elphick-Star** [comment]
@BenjaminAbt 
Thank you for the info and even more so for how quick you replied :-).

**Comment by Gary-Elphick-Star** [comment]
To follow that, I've just seen we were using the wrong authentication type on our SqlAzureDacpacDeployment@1 tasks, which it looks like we link to AAD.
https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/sql-azure-dacpac-deployment-v1?view=azure-pipelines

![image](https://github.com/dotnet/efcore/assets/146345692/8ac680f5-2316-43a1-b8e5-23194e4408ee)
